subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -96,13 +122,206 @@ is_target_video_sink (GstElement *element)
       g_strstr_len (klass, -1, "Video") && g_strstr_len (klass, -1, "Sink");
 }
 
+static GSocketAddress *
+socket_address_from_string (const char *name)
+{
+  int i, len;
+  static const char *unix_socket_address_types[] = {
+    "invalid",
+    "anonymous",
+    "path",
+    "abstract",
+    "padded"
+  };
+
+  for (i = 0; i < G_N_ELEMENTS (unix_socket_address_types); i++) {
+    len = strlen (unix_socket_address_types[i]);
+    if (!strncmp (name, unix_socket_address_types[i], len) && name[len] == ':') {
+      return g_unix_socket_address_new_with_type (name + len + 1, -1,
+          (GUnixSocketAddressType) i);
+    }
+  }
+
+  return NULL;
+}
+
+static GSocket *
+gst_framerate_client_socket_open (GstFramerateTracer *self)
+{
+  GSocket *socket;
+  GError *error = NULL;
+
+  socket = g_socket_new (G_SOCKET_FAMILY_UNIX, G_SOCKET_TYPE_STREAM, 0, &error);
+  if (socket == NULL) {
+    GST_ERROR_OBJECT (self, "%s", error->message);
+    g_error_free (error);
+    return NULL;
+  }
+  GST_DEBUG_OBJECT (self, "Socket created.");
+
+  g_socket_set_blocking (socket, FALSE);
+
+  return socket;
+}
+
+static void
+gst_framerate_client_socket_close (GstFramerateTracer *self, GSocket *socket)
+{
+  GError *error = NULL;
+
+  g_return_if_fail (socket != NULL);
+
+  if (!g_socket_close (socket, &error)) {
+    GST_ERROR_OBJECT (self, "Error closing socket: %s", error->message);
+  }
+  g_object_unref (socket);
+  if (error)
+    g_error_free (error);
+}
+
+static GIOStream *
+gst_framerate_client_socket_connect (GstFramerateTracer *self, GSocket *socket)
+{
+  GSocketAddress *addr;
+  GSocketConnectable *connectable;
+  GSocketAddressEnumerator *enumerator;
+  GSocketAddress *address;
+  GIOStream *connection;
+  GCancellable *cancellable = NULL;
+  GError *error = NULL;
+
+  g_return_val_if_fail (socket != NULL, NULL);
+
+  addr = socket_address_from_string (self->unix_socket_path);
+  if (addr == NULL) {
+    GST_ERROR_OBJECT (self, "Could not create a socket address at '%s'",
+        self->unix_socket_path);
+    return NULL;
+  }
+
+  connectable = G_SOCKET_CONNECTABLE (addr);
+  enumerator = g_socket_connectable_enumerate (connectable);
+  g_object_unref (connectable);
+
+  while (TRUE) {
+    address =
+        g_socket_address_enumerator_next (enumerator, cancellable, &error);
+    if (address == NULL) {
+      if (error == NULL)
+        GST_DEBUG_OBJECT (self, "No more addresses to try");
+      else {
+        GST_DEBUG_OBJECT (self, "%s", error->message);
+        g_error_free (error);
+      }
+      return NULL;
+    }
+
+    if (g_socket_connect (socket, address, cancellable, &error)) {
+      g_object_unref (address);
+      break;
+    }
+
+    GST_ERROR_OBJECT (self, "Connection to %s failed: %s, trying next",
+        g_unix_socket_address_get_path (G_UNIX_SOCKET_ADDRESS (address)),
+        error->message);
+    g_error_free (error);
+    error = NULL;
```

Comment:
`g_clear_pointer (&error, g_error_free);` will free the error if `!= NULL` and set it to NULL.